README
Hi!

You've most likely downloaded this package as a tar.gz distribution,
so you'll want to read Getting Started. If you are interested in
developing or submitting patches to the project, read the Contributing
and Coding Style sections.


Getting Started
---------------

You can grab a released version of Gearman from https://github.com/gearman/gearmand/. If you want to work on the latest code, please read the file HACKING.

To build, you can follow the normal:

./configure
make
make install

You can also run 'make test' before installing to make sure everything
checks out ok. Once you have it installed, you can start the Gearman
job server with:

gearmand -v

This will start it while printing some verbose messages. To try
running a job through it, look in the examples/ directory of this
source and run:

./reverse_worker

Once that is running, you can run your first job with:

./reverse_client "Hello, Gearman!"

If all goes well, the reverse_worker application should have output:

Job=H:lap:1 Workload=Hello, Gearman! Result=!namraeG ,olleH

While the reverse_client returned:

Result=!namraeG ,olleH

If you want to start writing your own client and workers, be sure to
check out the developer API at:

http://gearman.info/libgearman.html

You can also find other useful resources related to the project at:

http://www.gearman.org/

Enjoy!


Contributing
------------

If you are getting this code from https://github.com/gearman/gearmand
then continue reading. Otherwise these directions are not for you
(well maybe...).

To obtain code from Github you will need to
issue the following command:

git clone https://github.com/gearman/gearmand

Once the tree is branched you will need to build the "configure"
script. You can do this by running the script:

./bootstrap.sh -a

It will set up all of the files you need to build the package. At
that point it is just the typical "./configure; make; make test;
make install"

For a tarball release do a "make dist" and for an RPM type "make rpm".

For patches please create a branch on github and propose it to be
merged into the trunk. You can find out more information on how to
do this at the github help site:

https://help.github.com/articles/creating-a-pull-request/


Thanks and keep hacking!

Cheers,
  -Brian
  Seattle, WA.


ChangeLog
This is the hand-created ChangeLog used prior to v1.1.3. Please refer to git
log for changes after that.

1.1.12 Sun Feb  9 04:27:38 PST 2014
* GEARMAN_SERVERS environmental variable for libgearman to pick up servers to
communicate with. This means that any driver now linked with libgearman will
be able to handle multiple servers.
* Add INFO level messages for queue creation.

1.1.11 Thu Oct  3 04:38:47 EDT 2013
* Workers which return a bad gearman_return_t will be counted as an error by the server and not a final (i.e. they will be retried).
* Fixed possible bug where the server would over count the number of NOOP sent if NOOP messages did not get sent.

1.1.10 Mon Sep 16 04:20:13 CDT 2013
* Added gearman_job_use_client()
* Improve compile time.
* Fix for NOOP failure (bad worker causes early exit of loop).
* Fix for postgres (use INFORMATION_SCHEMA).
* Added gearman_client_has_active_tasks() so that you can see if a client has active tasks that it is working on.

1.1.9 Fri Aug  2 02:39:25 EDT 2013
* Added gearman_task_is_finished()
* Improved SSL support.
* Exceptions are now supported.
* gearmand accepts its root CA via the environmental variable GEARMAND_PORT. 
* libgearman will now accept GEARMAND_CA_CERTIFICATE, GEARMAN_CLIENT_PEM, and GEARMAN_CLIENT_KEY.

1.1.8 Thu Jun  6 18:47:01 EDT 2013
* Postgres test case now passes.
* SSL support added.
* OSX fixes.

1.1.7 Mon May  6 06:46:20 EDT 2013
* Cleanup of error codes returned by gearmand.
* gearmand will now set its port from the env variable GEARMAND_PORT.
* Fix issue where identifier might not be set correctly on reconnect.

1.1.6 Tue Apr 16 03:29:57 EDT 2013
* Merge of 1.0.4 tree
* Added support for gearadmin to "cancel" a job.
* Keep-alive support for gearmand has been extended (more options to control behavior).
* Fixed issues related to clients who didn't really support exceptions being passed exceptions.

1.1.5 Mon Feb  4 00:59:19 EST 2013
* Rollup of bug fixes for 1.0.3
* --threads=0 for gearmand will now result in gearmand using all available cores.

1.1.4 Mon Dec 17 21:24:16 EST 2012
* Add GEARMAN_CLIENT_GENERATE_UNIQUE, with default set to not generate.
* Experimental addition to queue service which will allow a queue to be stored on shutdown (--libsqlite3-store-on-shutdown).
* Rollup of all changes in 1.0.2

1.1.3 Wed Nov  7 22:48:21 EST 2012
* Merge with 1.0.1

1.1.2 Fri Oct 12 05:34:29 EDT 2012
* Merge with 0.41

1.1.1 Wed Sep 19 22:04:56 EDT 2012
* Merge with 0.39

1.1.0 Wed Sep  5 08:33:37 PDT 2012
* Fix for ABI compatibility issues.

1.0.6 Mon May  6 02:48:30 EDT 2013
* strlen() of zero on name passed for instance did not result in a connection to localhost (which the php driver thought would happen).
* Fix regression in log not updating time on all platforms.
* Fix issue where pecl php wasn't passing a null string to get default host.

1.0.5 Mon Apr 15 06:36:05 EDT 2013
* Fix for log_fn

1.0.4 Never Released
* Added --coredump flag to gearmand
* Removed result type enum, since the structure is never exposed to the user this was of limited use.
* Assorted bug fixes.

1.0.3 Sun Feb  3 21:05:52 EST 2013
* Allow for a longer set of retries if port is in use when the server starts up.
* Added checking for valgrind sgcheck
* Fedora 18 fixes.

1.0.2 Mon Dec 17 07:24:42 EST 2012
* Add --expceptions to Gearmand to enable exceptions by default
* Backtraces on some types of crashes.
* Add GEARMAN_CLIENT_GENERATE_UNIQUE.
* If the unique value passed to any client function is "#" then the unique value will be derived by hashing the workload that is being supplied to the function.
* Optimize status requests so that they exit as soon as request is received.
* Fix test case results for FreeBSD.

1.0.1 Tue Nov  6 01:07:07 EST 2012
* Stable branch labeled as 1.0
* Update all build related files.
* Add a few additional platforms to the supported list.

0.41 Fri Oct 12 05:34:29 EDT 2012
* Switched from --having-debug to --enable-debug
* Fix for SQLITE, it wasn't replaying its queue on startup.
* More testing done now on durable queues.

0.39 Wed Sep 19 03:36:06 EDT 2012
* Added gearman_job_error() interface.
* gearman_client_errno() and gearman_worker_errno() now return EINVAL if client or worker is invalid.

0.38 Fri Sep 14 02:16:01 EDT 2012
* Updates for compiling on different platforms.
* UUID portability fixes.

0.37 Sun Sep  2 22:51:08 EDT 2012
* Fix compiler issues around older compilers. Bug 1039629.

0.36 Sat Aug 18 16:57:13 EDT 2012
* Fix issue where sqlite was not correctly being shutdown.
* Check for more warnings (and cleanup found from).
* Fix thread safety issue for uuid_generate().

0.35 Sun Aug  5 16:03:21 EDT 2012
* Critical bug fix for rare case in which initialization of pipe for a worker fails.
* Add support for MySQL specific driver to specify port.
* Fix library issue with OSX Mountain Lion.

0.34 Tue Jul 10 21:19:10 EDT 2012
* Alpha support for direct MySQL queue.
* XML option for testing under valgrind.
* gearmand now allows for a user to specify the format for the server key
* First pass support for checking status via unique value.

0.33 Thu May  3 22:09:29 PDT 2012
* Add --config-file option
* Additional testing for Drizzle
* Fix for OSX 10.6.8 compile issue.

0.32 Wed Apr 11 02:25:38 PDT 2012
* Fix for shutdown bug.

0.31 Sat Apr  7 11:56:58 PDT 2012
* Fixes for bugs 975591, 974712, 970828, 967859, 967463,
* Valgrind fix around epoch retry.
* Numerous issues cleaned up around bouncing server during testing.

0.29 Tue Mar 13 22:58:16 PDT 2012
* Additional Fixes for Postgres
* Additional debug calls in server via tuneable logging.
* Documentation update.
* Fixes for Ubuntu 11.10

0.28 Thu Jan 26 22:43:46 PST 2012
* -v has been deprecated because of parsing issues. Please use --verbose instead
* --log-file=stderr will set output to stderr.
* Fix for gearman.spec file for rpm build.

0.27 Mon Jan  9 20:02:16 PST 2012
* Fix for TokyoCabinet
* Build updates for building on Ubuntu


util/daemon.h
/*    $Header: /cvsroot/wikipedia/willow/src/bin/willow/daemon.c,v 1.1 2005/05/02 19:15:21 kateturner Exp $    */
/*    $NetBSD: daemon.c,v 1.9 2003/08/07 16:42:46 agc Exp $    */
/*-
 * Copyright (c) 1990, 1993
 *    The Regents of the University of California.  All rights reserved.
 * Copyright (c) 2010
 *    Stewart Smith
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#pragma once

namespace gearmand
{

bool daemon_is_ready(bool close_io);
bool daemonize(bool is_chdir= true, bool wait_sigusr1= true);

}
